### PR TITLE
Fix to service name and link name validation to allow single character

### PR DESCRIPTION
**Pull request by prachidamle**
Fix to service name and link name validation to allow single character in the name
Also, added the fix for cross stack links which can contain /
Added some tests too

https://github.com/rancher/rancher/issues/4090

### CODE REVIEW

**Comment by prachidamle** [comment]
@ibuildthecloud @vincent99 this is to fix the single char bug in the regex used

**Comment by vincent99** [comment]
I suppose this works, probably, but I've done a lot of regex and don't know exactly what `(?!.*--)` means without looking it up, and then you're adding basic string checks because the regex doesn't actually work.  So why not just make all the exceptions simple checks and us a human-readable regex:

```
if ( matches('[^a-zA-Z0-9-]') || startsWith('-') || endsWith('-') || contains('--') {
  // bad
}
else {
  // good
}
```

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/filter/common/AbstractDefaultResourceManagerFilter.java [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 public class AbstractDefaultResourceManagerFilter extends AbstractResourceManagerFilter implements Priority {
 
-    private static final Pattern DNS_NAME_PATTERN = Pattern.compile("^[a-zA-Z0-9](?!.*--)[a-zA-Z0-9-]*[a-zA-Z0-9]$");
+    private static final Pattern DNS_NAME_PATTERN = Pattern.compile("^[a-zA-Z0-9](?!.*--)[a-zA-Z0-9-]*[a-zA-Z0-9]*$");
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/filter/common/AbstractDefaultResourceManagerFilter.java]: The end of this pattern is redundant.

Comment (prachidamle) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/filter/common/AbstractDefaultResourceManagerFilter.java]: @vincent99 (?!{regex}) is the negative lookahead supported by java to match "something not followed by something" pattern

Comment (prachidamle) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/filter/common/AbstractDefaultResourceManagerFilter.java]: @ibuildthecloud I have removed the redundant portion.

**Pull request merged**